spices/SPICE-0009-external-readers.adoc

```diff
@@ -0,0 +1,319 @@
+
+The existing message passing API and client libraries will be expanded to support external readers.
+Instead of client libraries invoking `pkl server`, `pkl eval` invocations, when configured with an external reader, will run the reader executable as a subprocess.
+External readers can provide any number of resource and/or module schemes and will be configured as an executable and list of arguments via:
+
+* CLI flag
+* PklProject property
+* Java API
+* Message passing API
+
+[source,pkl]
+----
+/// May be specified as an absolute path to an executable
+/// May also be specified as just an executable name, in which case it will be resolved according to the PATH environment variable
+executable: String
+
+/// Command line arguments that will be passed to the reader process
+arguments: Listing<String>
+----
+
+=== Example
+
+Consider this module:
+
+[source,pkl]
+----
+username = "john_appleseed"
+
+email = read("ldap://ds.example.com:389/dc=example,dc=com?mail?sub?(uid=\(username))").text
+----
+
+Pkl doesn't implement the `ldap:` resource URI scheme natively, but an external reader can provide it.
+Assuming a hypothetical `pkl-ldap` binary implementing the external reader protocol and the `ldap:` scheme is in the `$PATH`, this module can be evaluated as:
+
+[source,text]
+----
+$ pkl eval <module> --external-reader=pkl-ldap --allowed-resources prop:,ldap:
+username = "john_appleseed"
+email = "appleseed@example.com"
+----
+
+For security reasons, schemes provided by external readers are not allowed by default and clients must explicitly allow the desired scheme(s).
+In this example, the external reader may provide both `ldap:` and `ldaps:` schemes.
+If a codebase or user wishes to enforce that only secure LDAP connections are made, the `ldaps:` scheme may be allowed while `ldap:` is not.
+
+== Detailed design
+
+To avoid terminology confusion with the existing language binding message passing model, the `pkl` process will remain known as the "server" and the reader process will be known as the "client", despite all requests in this proposal being initiated by the `pkl` process.
+
+=== Overall Flow
+
+1. User requests an evaluator (via Java API, Message Passing API, or CLI) with an external reader.
+2. Entrypoint (client Java code, `CliEvaluator`, or `Server`) instantiates an `ExternalReader`.
+3. The `ExternalReader` is started, which starts the configured child process and sends a `DiscoverReadersRequest` message.
+4. A `DiscoverReadersResponse` message is received (within some timeout).
+5. The entrypoint requests resource readers (`ExternalResourceReader`) and module key factories (`ExternalModuleKeyFactory`) from the `ExternalReader` and uses them to build an evaluator.
+`ExternalResourceReader` and `ExternalModuleKeyFactory` function similarly to the existing `ClientResourceReader` and `ClientModuleKeyFactory`, implementation may be reusable entirely.
+6. Evaluation begins.
+7. A resource or module read is encountered.
```

Comment:
I think steps 2 - 5 should be inserted here. We should defer spawning a child process to when it's actually needed.

Reply:
So that I'm sure I'm understanding correctly, instead of declaring the scheme->command process up-front, instead, the first time the `ExternalModuleKeyFactory.create` is called, launch the process and perform the discovery. That would be really ideal, but the challenge I'm seeing is that `ResourceReader` must implement `getUriScheme` so the schemes must be known at `ResourceManager` init time. Am I understanding you correctly? If so, any recommendations on how to handle this for resource readers?

Reply:
Sorry, I'm simply saying that we should defer spawning a process to when that process is actually needed, rather than spawn it prior to the start of evaluation. That way, Pkl only pays the cost of spawning that process for programs that actually read from a resource.